File: leetcode/merge_intervals/merge_intervals.cpp
#include <vector>
#include <algorithm>

using std::vector;
using std::sort;
using std::max;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if(intervals.size() < 2) return intervals;
        sort(intervals.begin(), intervals.end());

        vector<vector<int>> ans;
        ans.push_back(intervals[0]);

        int i = 1, maxInterval;
        while(i < intervals.size()) {
            if(intervals[i][0] > ans.back()[1]) {
                ans.push_back(intervals[i++]);
            } else {
                maxInterval = ans.back()[1];
                while(i < intervals.size() && intervals[i][0] <= maxInterval) maxInterval = max(maxInterval, intervals[i++][1]);
                ans.back()[1] = maxInterval;
            }
        }

        return ans;
    }
};

File: leetcode/merge_intervals/merge_intervals.java
package merge_intervals;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;

class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals.length < 2)
            return intervals;

        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));

        ArrayList<int[]> ans = new ArrayList<>();
        ans.add(intervals[0]);
        int i = 0, maxInterval, lastIndex;

        while (i < intervals.length) {
            lastIndex = ans.size() - 1;
            if (intervals[i][0] > ans.get(lastIndex)[1]) {
                ans.add(intervals[i++]);
            } else {
                maxInterval = ans.get(lastIndex)[1];
                while (i < intervals.length && intervals[i][0] <= maxInterval) maxInterval = Math.max(maxInterval, intervals[i++][1]);
                ans.get(lastIndex)[1] = maxInterval;
            }
        }
        return ans.toArray(new int[0][]);
    }
}
